### ISSUE TITLE
Recover from `impl &Trait` and `T: &Trait`

### ISSUE THREAD

**Issue by estebank** [rust-lang/rust#106694]
[The following two function definitions are not valid](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9bda53271ef3a8886793cf427b8cea91):
```rust
fn foo(_: impl &Trait) {}
fn bar<T: &Trait>(_: T) {}
```

They fail in the parser and it doesn't recover nicely.

```
error: expected one of `:`, `@`, or `|`, found `)`
 --> src/main.rs:2:22
  |
2 | fn foo(_: impl &Trait) {}
  |                      ^ expected one of `:`, `@`, or `|`
  |
  = note: anonymous parameters are removed in the 2018 edition (see RFC 1685)
help: if this is a parameter name, give it a type
  |
2 | fn foo(_: impl Trait: &TypeName) {}
  |                ~~~~~~~~~~~~~~~~
help: if this is a type, explicitly ignore the parameter name
  |
2 | fn foo(_: impl _: &Trait) {}
  |                ++

error: expected one of `!`, `(`, `)`, `,`, `?`, `for`, `~`, lifetime, or path, found `&`
 --> src/main.rs:2:16
  |
2 | fn foo(_: impl &Trait) {}
  |               -^ expected one of 9 possible tokens
  |               |
  |               help: missing `,`

error: expected one of `!`, `(`, `,`, `=`, `>`, `?`, `for`, `~`, lifetime, or path, found `&`
 --> src/main.rs:3:11
  |
3 | fn bar<T: &Trait>(_: T) {}
  |           ^ expected one of 10 possible tokens
```

It should produce an appropriate error message, closer to the output for `impl &Trait for () {}`:

```
error: expected a trait, found type
 --> src/main.rs:2:6
  |
2 | impl &Trait for () {}
  |      ^^^^^^
```

**Comment by Ezrashaw** [comment]
@rustbot claim

Might take a while, it is P-low after all.
